Makefile

```diff
@@ -39,48 +39,61 @@ help:
 setup: .env .venv requirements.txt.log requirements-mysql.txt.log requirements-postgresql.txt.log  ## Sets up the project (installs dependencies etc.)
 
 
-requirements.txt.log: requirements.txt
-	@pip install -r requirements.txt | tee .requirements.txt.tmp.log
+requirements.txt.log: requirements.txt.freeze
+	@. .venv/bin/activate && pip install -r requirements.txt.freeze | tee .requirements.txt.tmp.log
 	@mv .requirements.txt.tmp.log requirements.txt.log
 
-requirements-mysql.txt.log: .env requirements-mysql.txt
-	@(grep '^W2_DATABASE_ENGINE' .env | grep 'mysql' > /dev/null && pip install -r requirements-mysql.txt || echo "Not using MySQL") | tee .requirements-mysql.txt.tmp.log
+requirements-mysql.txt.log: .env requirements-mysql.txt.freeze
+	@. .venv/bin/activate && (grep '^W2_DATABASE_ENGINE' .env | grep 'mysql' > /dev/null && pip install -r requirements-mysql.txt.freeze || echo "Not using MySQL") | tee .requirements-mysql.txt.tmp.log
 	@mv .requirements-mysql.txt.tmp.log requirements-mysql.txt.log
 
-requirements-postgresql.txt.log: .env requirements-postgresql.txt
-	@(grep '^W2_DATABASE_ENGINE' .env | grep 'postgresql' > /dev/null && pip install -r requirements-postgresql.txt || echo "Not using PostgreSQL") | tee .requirements-postgresql.txt.tmp.log
+requirements-postgresql.txt.log: .env requirements-postgresql.txt.freeze
+	@. .venv/bin/activate && (grep '^W2_DATABASE_ENGINE' .env | grep 'postgresql' > /dev/null && pip install -r requirements-postgresql.txt.freeze || echo "Not using PostgreSQL") | tee .requirements-postgresql.txt.tmp.log
 	@mv .requirements-postgresql.txt.tmp.log requirements-postgresql.txt.log
 
 
+.PHONY: freeze-dependencies
+freeze-dependencies:
+	@. .venv/bin/activate && pip freeze | grep -v 'mysqlclient' > requirements.txt.freeze
+	@. .venv/bin/activate && (grep '^W2_DATABASE_ENGINE' .env | grep 'mysql' > /dev/null && (pip freeze | grep 'mysqlclient' > requirements-mysql.txt.freeze || true) || true)
+	@. .venv/bin/activate && (grep '^W2_DATABASE_ENGINE' .env | grep 'postgresql' > /dev/null && (pip freeze | grep 'psycopg2' > requirements-postgres.txt.freeze || true) || true)
+
+
+.PHONY: update-dependencies
+update-dependencies:
+	@rm -rf .venv/
+	@make .venv
+	@. .venv/bin/activate && pip install -r requirements.txt
+	@make freeze-dependencies
+
+
 .PHONY: test
 test: setup  ## Runs the unit tests
-	@. .venv/bin/activate
-	@./manage.py test
+	@. .venv/bin/activate && ./manage.py test
```

Comment:
Oh yeah.. While I was in the area - I fixed a bug.. I didn't notice this when developing because I had just been installing these packages into my global python package list :facepalm: .
Thing is - the sourcing of the `activate` file doesn't persist into the next line in the target recipe.